core/src/main/java/org/apache/iceberg/rest/RESTContentFileParser.java

```diff
@@ -0,0 +1,250 @@
+  private static final String EQUALITY_IDS = "equality-ids";
+  private static final String SORT_ORDER_ID = "sort-order-id";
+
+  private RESTContentFileParser() {}
+
+  public static String toJson(ContentFile<?> contentFile) {
+    return JsonUtil.generate(
+        generator -> RESTContentFileParser.toJson(contentFile, generator), false);
+  }
+
+  public static void toJson(ContentFile<?> contentFile, JsonGenerator generator)
+      throws IOException {
+    Preconditions.checkArgument(contentFile != null, "Invalid content file: null");
+    Preconditions.checkArgument(generator != null, "Invalid JSON generator: null");
+
+    generator.writeStartObject();
+
+    generator.writeNumberField(SPEC_ID, contentFile.specId());
+    generator.writeStringField(CONTENT, contentFile.content().name());
+    generator.writeStringField(FILE_PATH, contentFile.path().toString());
+    generator.writeStringField(FILE_FORMAT, contentFile.format().name());
+
+    generator.writeFieldName(PARTITION);
+
+    // TODO at the time of serialization we dont have the partition spec we just have spec id.
+    // we will need to get the spec from table metadata using spec id.
+    // or we will need to send parition spec, put null here for now until refresh
+    SingleValueParser.toJson(null, contentFile.partition(), generator);
+
+    generator.writeNumberField(FILE_SIZE_IN_BYTES, contentFile.fileSizeInBytes());
+
+    metricsToJson(contentFile, generator);
+
+    if (contentFile.keyMetadata() != null) {
+      generator.writeFieldName(KEY_METADATA);
+      SingleValueParser.toJson(DataFile.KEY_METADATA.type(), contentFile.keyMetadata(), generator);
+    }
+
+    if (contentFile.splitOffsets() != null) {
+      JsonUtil.writeLongArray(SPLIT_OFFSETS, contentFile.splitOffsets(), generator);
+    }
+
+    if (contentFile.equalityFieldIds() != null) {
+      JsonUtil.writeIntegerArray(EQUALITY_IDS, contentFile.equalityFieldIds(), generator);
+    }
+
+    if (contentFile.sortOrderId() != null) {
+      generator.writeNumberField(SORT_ORDER_ID, contentFile.sortOrderId());
+    }
+
+    generator.writeEndObject();
+  }
+
+  public static ContentFile<?> fromJson(JsonNode jsonNode) {
+    Preconditions.checkArgument(jsonNode != null, "Invalid JSON node for content file: null");
+    Preconditions.checkArgument(
+        jsonNode.isObject(), "Invalid JSON node for content file: non-object (%s)", jsonNode);
+
+    int specId = JsonUtil.getInt(SPEC_ID, jsonNode);
+    FileContent fileContent = FileContent.valueOf(JsonUtil.getString(CONTENT, jsonNode));
+    String filePath = JsonUtil.getString(FILE_PATH, jsonNode);
+    FileFormat fileFormat = FileFormat.fromString(JsonUtil.getString(FILE_FORMAT, jsonNode));
+
+    // TODO at the time of deserialization we dont have the partition spec we just have spec id.
+    // we will need to get the spec from table metadata using spec id.
+    // we wil need to send parition spec i believe, put some placeholder here for now null
```

Comment:
Yeah I think there's a compelling case to update the rest spec so we send back the full serialized spec .In case we were worried about payload sizes going over the wire we could follow the same concept we do for delete file references where there's a top level list where it's only listed once, and in the tasks itself it's just an integer index. But this complicates the protocol.

 I think payload size is more of a concern for sending schemas, which I'm pretty against. I feel like resolving schemas is probably worth it on the client. Besides clients have full context on what's being projected and can just send the right schema to the file scan task. There's also an argument for the server sending back a schema ID so that the client doesn't have to do the work of resolving a schema to use for passing to the file scan task, but the schema in FileScanTask is ultimately for any client side grouping the client wants to do and we could just use the current schema or the snapshot schema.

Reply:
> we send back the full serialized partition spec .In case we were worried about payload sizes going over the wire we could follow the same concept we do for delete file references where there's a top level list where it's only listed once, and in the tasks itself it's just an integer index. But this complicates the protocol. 

presently this leaves us with no option but to bind the partition spec with the current schema which is tbh not a bad option as schema / partition evolution cases takes care of not allowing backward incompatible schema by default. interesting would be cases when we allow this flag in schema evolution ? 

Having schema_id would be benefitial tbh as if for now we want to project col a, col b and the data file not having any of them would still be returned to the client which eventually the engines / caller will discard but this makes me think more in terms of column level access control, but totally understand we lack this mechanism today.